### ISSUE TITLE
Incorrect local threadgroup sizes in compute shader when using specialization info

### ISSUE THREAD

**Issue by mbarriault** [KhronosGroup/MoltenVK#188]
I have a compute shader that acts on 1-dimensional data. I declare the threadgroup size with specializations as (in GLSL)

```
layout(local_size_x_id = 0, local_size_y = 1, local_size_z = 1)
```

The y and z are, to my understanding, optional, but included in my testing for brevity. As far as I know as well, this is fine in SPIR-V, and checking the output verifies this. I bind this during pipeline creation

```
static const vk::SpecializationMapEntry mapEntry(0, 0, sizeof(uint32_t));
static const vk::SpecializationInfo specializationInfo(1, &mapEntry, sizeof(uint32_t), reinterpret_cast<const void*>(&localSize));
pipelineInfo.stage.pSpecializationInfo = &specializationInfo;
```

However, when I run, the threadgroup has size `(localSize, localSize, localSize)` (which crashes because `localSize*localSize*localSize` is larger than the threadgroup size limit), instead of the expected `(localSize, 1, 1)`.

As a workaround currently, I specify the other two (redundant) components,

```
layout(local_size_x_id = 0, local_size_y_id = 1, local_size_z_id = 2)
```

and I map

```
static const uint32_t localSizes[] = {localSize, 1, 1};
static const vk::SpecializationMapEntry mapEntries[] = {
    {0, 0*sizeof(uint32_t), sizeof(uint32_t)},
    {1, 1*sizeof(uint32_t), sizeof(uint32_t)},
    {2, 2*sizeof(uint32_t), sizeof(uint32_t)}
};
static const vk::SpecializationInfo specializationInfo(3, mapEntries, 3*sizeof(uint32_t), reinterpret_cast<const void*>(localSizes));
pipelineInfo.stage.pSpecializationInfo = &specializationInfo;
```

then I get the expected threadgroup size of `(localSize, 1, 1)`.

**Comment by billhollings** [comment]
@mbarriault 

Can you log the MSL shader compilation, and post the SPIR-V and resulting MSL here, please?

You can do this by turning on `debugMode` within MoltenVK, by either compiling MoltenVK in Debug mode in Xcode...or by using `vkGetMoltenVKDeviceConfigurationMVK()` and `vkSetMoltenVKDeviceConfigurationMVK()` to set MoltenVK configuration info in the MoltenVK runtime.

The Hologram demo app in the MoltenVK distribution includes an example of this last step;

    #include <MoltenVK/vk_mvk_moltenvk.h>

    MVKDeviceConfiguration mvkConfig;
    vkGetMoltenVKDeviceConfigurationMVK(vkDev, &mvkConfig);
    mvkConfig.debugMode = true;
    vkSetMoltenVKDeviceConfigurationMVK(vkDev, &mvkConfig);

**Comment by mbarriault** [comment]
The input GLSL

```
Read GLSL from file: particles.comp
Converting GLSL:
#version 450

layout(local_size_x_id = 0, local_size_y = 1, local_size_z = 1) in;

layout(binding = 0) buffer Input {
    mat4 model[];
};
layout(binding = 1) buffer Timer {
    uint timer;
};
layout(binding = 2) buffer Valid {
    uint alive[];
};
layout(binding = 3) buffer Noise {
    float noise[];
};

const float chance = 0.9995;

void main() {
    uint id = gl_GlobalInvocationID.x;
    const float dt = 1.0 / 60.0;

    if (id == 0)
        atomicAdd(timer, 1);

    barrier();

    float t = float(timer);

    float c = noise[id];
    if (alive[id] == 0) {
        if (c > chance) {
            float a = (c - chance) / (1 - chance) * 2*acos(-1);
            model[id][0].x = cos(a);
            model[id][0].y = sin(a);
            model[id][1].x = -sin(a);
            model[id][1].y = cos(a);
            model[id][3].xyz = vec3(0);
            alive[id] = 1;
        }
    }
    else /*if (alive[id] == 1)*/ {
        mat3 R = mat3(model[id]);
        model[id][3].xyz += dt * model[id][0].xyz;

        if (length(model[id][3].xyz) > 1 - c/1.5)
        {
            alive[id] = 0;
            model[id][3].xyz = vec3(0);
        }
    }
}
```

The converted SPIR-V:

```
; SPIR-V
; Version: 1.0
; Generator: Khronos Glslang Reference Front End; 7
; Bound: 162
; Schema: 0
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %4 "main" %11
               OpExecutionMode %4 LocalSize 1 1 1
               OpSource GLSL 450
               OpName %4 "main"
               OpName %8 "id"
               OpName %11 "gl_GlobalInvocationID"
               OpName %21 "Timer"
               OpMemberName %21 0 "timer"
               OpName %23 ""
               OpName %34 "t"
               OpName %38 "c"
               OpName %40 "Noise"
               OpMemberName %40 0 "noise"
               OpName %42 ""
               OpName %48 "Valid"
               OpMemberName %48 0 "alive"
               OpName %50 ""
               OpName %62 "a"
               OpName %74 "Input"
               OpMemberName %74 0 "model"
               OpName %76 ""
               OpName %109 "R"
               OpDecorate %11 BuiltIn GlobalInvocationId
               OpMemberDecorate %21 0 Offset 0
               OpDecorate %21 BufferBlock
               OpDecorate %23 DescriptorSet 0
               OpDecorate %23 Binding 1
               OpDecorate %39 ArrayStride 4
               OpMemberDecorate %40 0 Offset 0
               OpDecorate %40 BufferBlock
               OpDecorate %42 DescriptorSet 0
               OpDecorate %42 Binding 3
               OpDecorate %47 ArrayStride 4
               OpMemberDecorate %48 0 Offset 0
               OpDecorate %48 BufferBlock
               OpDecorate %50 DescriptorSet 0
               OpDecorate %50 Binding 2
               OpDecorate %73 ArrayStride 64
               OpMemberDecorate %74 0 ColMajor
               OpMemberDecorate %74 0 Offset 0
               OpMemberDecorate %74 0 MatrixStride 16
               OpDecorate %74 BufferBlock
               OpDecorate %76 DescriptorSet 0
               OpDecorate %76 Binding 0
               OpDecorate %160 SpecId 0
               OpDecorate %161 BuiltIn WorkgroupSize
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeInt 32 0
          %7 = OpTypePointer Function %6
          %9 = OpTypeVector %6 3
         %10 = OpTypePointer Input %9
         %11 = OpVariable %10 Input
         %12 = OpConstant %6 0
         %13 = OpTypePointer Input %6
         %17 = OpTypeBool
         %21 = OpTypeStruct %6
         %22 = OpTypePointer Uniform %21
         %23 = OpVariable %22 Uniform
         %24 = OpTypeInt 32 1
         %25 = OpConstant %24 0
         %26 = OpTypePointer Uniform %6
         %28 = OpConstant %6 1
         %30 = OpConstant %6 2
         %31 = OpConstant %6 264
         %32 = OpTypeFloat 32
         %33 = OpTypePointer Function %32
         %39 = OpTypeRuntimeArray %32
         %40 = OpTypeStruct %39
         %41 = OpTypePointer Uniform %40
         %42 = OpVariable %41 Uniform
         %44 = OpTypePointer Uniform %32
         %47 = OpTypeRuntimeArray %6
         %48 = OpTypeStruct %47
         %49 = OpTypePointer Uniform %48
         %50 = OpVariable %49 Uniform
         %58 = OpConstant %32 0.999499977
         %65 = OpConstant %32 0.000500000024
         %67 = OpConstant %32 2
         %69 = OpConstant %32 3.14159274
         %71 = OpTypeVector %32 4
         %72 = OpTypeMatrix %71 4
         %73 = OpTypeRuntimeArray %72
         %74 = OpTypeStruct %73
         %75 = OpTypePointer Uniform %74
         %76 = OpVariable %75 Uniform
         %86 = OpConstant %24 1
         %96 = OpConstant %24 3
         %97 = OpTypeVector %32 3
         %98 = OpConstant %32 0
         %99 = OpConstantComposite %97 %98 %98 %98
        %100 = OpTypePointer Uniform %71
        %107 = OpTypeMatrix %97 3
        %108 = OpTypePointer Function %107
        %111 = OpTypePointer Uniform %72
        %114 = OpConstant %32 1
        %129 = OpConstant %32 0.0166666675
        %148 = OpConstant %32 1.5
        %160 = OpSpecConstant %6 1
        %161 = OpSpecConstantComposite %9 %160 %28 %28
          %4 = OpFunction %2 None %3
          %5 = OpLabel
          %8 = OpVariable %7 Function
         %34 = OpVariable %33 Function
         %38 = OpVariable %33 Function
         %62 = OpVariable %33 Function
        %109 = OpVariable %108 Function
         %14 = OpAccessChain %13 %11 %12
         %15 = OpLoad %6 %14
               OpStore %8 %15
         %16 = OpLoad %6 %8
         %18 = OpIEqual %17 %16 %12
               OpSelectionMerge %20 None
               OpBranchConditional %18 %19 %20
         %19 = OpLabel
         %27 = OpAccessChain %26 %23 %25
         %29 = OpAtomicIAdd %6 %27 %28 %12 %28
               OpBranch %20
         %20 = OpLabel
               OpControlBarrier %30 %30 %31
         %35 = OpAccessChain %26 %23 %25
         %36 = OpLoad %6 %35
         %37 = OpConvertUToF %32 %36
               OpStore %34 %37
         %43 = OpLoad %6 %8
         %45 = OpAccessChain %44 %42 %25 %43
         %46 = OpLoad %32 %45
               OpStore %38 %46
         %51 = OpLoad %6 %8
         %52 = OpAccessChain %26 %50 %25 %51
         %53 = OpLoad %6 %52
         %54 = OpIEqual %17 %53 %12
               OpSelectionMerge %56 None
               OpBranchConditional %54 %55 %106
         %55 = OpLabel
         %57 = OpLoad %32 %38
         %59 = OpFOrdGreaterThan %17 %57 %58
               OpSelectionMerge %61 None
               OpBranchConditional %59 %60 %61
         %60 = OpLabel
         %63 = OpLoad %32 %38
         %64 = OpFSub %32 %63 %58
         %66 = OpFDiv %32 %64 %65
         %68 = OpFMul %32 %66 %67
         %70 = OpFMul %32 %68 %69
               OpStore %62 %70
         %77 = OpLoad %6 %8
         %78 = OpLoad %32 %62
         %79 = OpExtInst %32 %1 Cos %78
         %80 = OpAccessChain %44 %76 %25 %77 %25 %12
               OpStore %80 %79
         %81 = OpLoad %6 %8
         %82 = OpLoad %32 %62
         %83 = OpExtInst %32 %1 Sin %82
         %84 = OpAccessChain %44 %76 %25 %81 %25 %28
               OpStore %84 %83
         %85 = OpLoad %6 %8
         %87 = OpLoad %32 %62
         %88 = OpExtInst %32 %1 Sin %87
         %89 = OpFNegate %32 %88
         %90 = OpAccessChain %44 %76 %25 %85 %86 %12
               OpStore %90 %89
         %91 = OpLoad %6 %8
         %92 = OpLoad %32 %62
         %93 = OpExtInst %32 %1 Cos %92
         %94 = OpAccessChain %44 %76 %25 %91 %86 %28
               OpStore %94 %93
         %95 = OpLoad %6 %8
        %101 = OpAccessChain %100 %76 %25 %95 %96
        %102 = OpLoad %71 %101
        %103 = OpVectorShuffle %71 %102 %99 4 5 6 3
               OpStore %101 %103
        %104 = OpLoad %6 %8
        %105 = OpAccessChain %26 %50 %25 %104
               OpStore %105 %28
               OpBranch %61
         %61 = OpLabel
               OpBranch %56
        %106 = OpLabel
        %110 = OpLoad %6 %8
        %112 = OpAccessChain %111 %76 %25 %110
        %113 = OpLoad %72 %112
        %115 = OpCompositeExtract %32 %113 0 0
        %116 = OpCompositeExtract %32 %113 0 1
        %117 = OpCompositeExtract %32 %113 0 2
        %118 = OpCompositeExtract %32 %113 1 0
        %119 = OpCompositeExtract %32 %113 1 1
        %120 = OpCompositeExtract %32 %113 1 2
        %121 = OpCompositeExtract %32 %113 2 0
        %122 = OpCompositeExtract %32 %113 2 1
        %123 = OpCompositeExtract %32 %113 2 2
        %124 = OpCompositeConstruct %97 %115 %116 %117
        %125 = OpCompositeConstruct %97 %118 %119 %120
        %126 = OpCompositeConstruct %97 %121 %122 %123
        %127 = OpCompositeConstruct %107 %124 %125 %126
               OpStore %109 %127
        %128 = OpLoad %6 %8
        %130 = OpLoad %6 %8
        %131 = OpAccessChain %100 %76 %25 %130 %25
        %132 = OpLoad %71 %131
        %133 = OpVectorShuffle %97 %132 %132 0 1 2
        %134 = OpVectorTimesScalar %97 %133 %129
        %135 = OpAccessChain %100 %76 %25 %128 %96
        %136 = OpLoad %71 %135
        %137 = OpVectorShuffle %97 %136 %136 0 1 2
        %138 = OpFAdd %97 %137 %134
        %139 = OpAccessChain %100 %76 %25 %128 %96
        %140 = OpLoad %71 %139
        %141 = OpVectorShuffle %71 %140 %138 4 5 6 3
               OpStore %139 %141
        %142 = OpLoad %6 %8
        %143 = OpAccessChain %100 %76 %25 %142 %96
        %144 = OpLoad %71 %143
        %145 = OpVectorShuffle %97 %144 %144 0 1 2
        %146 = OpExtInst %32 %1 Length %145
        %147 = OpLoad %32 %38
        %149 = OpFDiv %32 %147 %148
        %150 = OpFSub %32 %114 %149
        %151 = OpFOrdGreaterThan %17 %146 %150
               OpSelectionMerge %153 None
               OpBranchConditional %151 %152 %153
        %152 = OpLabel
        %154 = OpLoad %6 %8
        %155 = OpAccessChain %26 %50 %25 %154
               OpStore %155 %12
        %156 = OpLoad %6 %8
        %157 = OpAccessChain %100 %76 %25 %156 %96
        %158 = OpLoad %71 %157
        %159 = OpVectorShuffle %71 %158 %99 4 5 6 3
               OpStore %157 %159
               OpBranch %153
        %153 = OpLabel
               OpBranch %56
         %56 = OpLabel
               OpReturn
               OpFunctionEnd
```

and the final MSL

```
#pragma clang diagnostic ignored "-Wunused-variable"

#include <metal_stdlib>
#include <simd/simd.h>
#include <metal_atomic>

using namespace metal;

constant uint _160_tmp [[function_constant(0)]];
constant uint _160 = is_function_constant_defined(_160_tmp) ? _160_tmp : 1u;
constant uint3 gl_WorkGroupSize = uint3(_160, 1u, 1u);

struct Timer
{
    uint timer;
};

struct Noise
{
    float noise[1];
};

struct Valid
{
    uint alive[1];
};

struct Input
{
    float4x4 model[1];
};

kernel void main0(device Input& _76 [[buffer(0)]], device Timer& _23 [[buffer(1)]], device Valid& _50 [[buffer(2)]], device Noise& _42 [[buffer(3)]], uint3 gl_GlobalInvocationID [[thread_position_in_grid]])
{
    uint id = gl_GlobalInvocationID.x;
    if (id == 0u)
    {
        uint _29 = atomic_fetch_add_explicit((volatile device atomic_uint*)&_23.timer, 1u, memory_order_relaxed);
    }
    threadgroup_barrier(mem_flags::mem_threadgroup);
    float t = float(_23.timer);
    float c = _42.noise[id];
    if (_50.alive[id] == 0u)
    {
        if (c > 0.9995)
        {
            float a = (((c - 0.9995) / 0.0005) * 2.0) * 3.14159;
            _76.model[id][0].x = cos(a);
            _76.model[id][0].y = sin(a);
            _76.model[id][1].x = -sin(a);
            _76.model[id][1].y = cos(a);
            _76.model[id][3] = float4(float3(0.0).x, float3(0.0).y, float3(0.0).z, _76.model[id][3].w);
            _50.alive[id] = 1u;
        }
    }
    else
    {
        float3x3 R = float3x3(float3(_76.model[id][0].x, _76.model[id][0].y, _76.model[id][0].z), float3(_76.model[id][1].x, _76.model[id][1].y, _76.model[id][1].z), float3(_76.model[id][2].x, _76.model[id][2].y, _76.model[id][2].z));
        float3 _138 = _76.model[id][3].xyz + (_76.model[id][0].xyz * 0.0166667);
        _76.model[id][3] = float4(_138.x, _138.y, _138.z, _76.model[id][3].w);
        if (length(_76.model[id][3].xyz) > (1.0 - (c / 1.5)))
        {
            _50.alive[id] = 0u;
            _76.model[id][3] = float4(float3(0.0).x, float3(0.0).y, float3(0.0).z, _76.model[id][3].w);
        }
    }
}
```

The money shot is in the SPIR-V

```
               OpDecorate %160 SpecId 0
               OpDecorate %161 BuiltIn WorkgroupSize
...
          %6 = OpTypeInt 32 0
         %28 = OpConstant %6 1
...
        %160 = OpSpecConstant %6 1
        %161 = OpSpecConstantComposite %9 %160 %28 %28
```

and the matching MSL

```
constant uint _160_tmp [[function_constant(0)]];
constant uint _160 = is_function_constant_defined(_160_tmp) ? _160_tmp : 1u;
constant uint3 gl_WorkGroupSize = uint3(_160, 1u, 1u);
```

The shader conversion is doing the right thing, mapping SPIR-V specializations to MSL function constants, but MoltenVK isn't interpreting it right, assigning `_160` to the other components of `gl_WorkGroupSize`.

**Comment by billhollings** [comment]
Fixed in PR #191. Please retest and close.

**Comment by mbarriault** [comment]
That seems to have done the trick. Using the original code as above works now as intended. Cheers!
